Generate documentation using Analysis Server, when possible.

scheglov opened the pull request:
This may be a solution for [WEB-15936](https://youtrack.jetbrains.com/issue/WEB-15936).

When DAS.getHover() is used, we can show the exact type information about the given element  both static and propagated. Note that even the static type is not always the same as the declared type because of the type promotion process or "strong mode".

alexander-doroshko commented:
Which one do you like more?
![image](https://cloud.githubusercontent.com/assets/4499789/11749059/a592e14a-a03c-11e5-95e8-2ed3b21e8b85.png)
For me the right one looks better. I see that Foo is a class and no senseless signature and type.

alexander-doroshko commented on Dart/testSrc/com/jetbrains/lang/dart/documentation/DartDocUtilTest.java:
```diff
@@ -22,58 +22,58 @@ private void doTest(String expectedDoc, String fileContents) {
 
   public void testAbstractClassSig() throws Exception {
-    doTest("<code><small><b>test.dart</b></small></code><br/><br/><code>abstract class <b>Foo</b><br/>extends Bar</code>",
+    doTest("<code><b>Containing library:</b> <code><small>test.dart</small></code><br/><br/><code>abstract class <b>Foo</b> extends Bar</code>",
```
Unbalanced <code> tags in expected data is suspicious

alexander-doroshko commented on Dart/src/com/jetbrains/lang/dart/analyzer/DartAnalysisServerService.java:
```diff
@@ -78,6 +79,8 @@
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.atomic.AtomicReferenceArray;
```
unused import

alexander-doroshko replied:
Hint: to avoid unused imports and other typos I always review the code before commit. I press Ctrl+D (Cmd+D) in Local Changes tab right before commit or right in the commit dialog and browse all changes by simply pressing F7. Editing and reformatting code works in the diff view.

alexander-doroshko commented on Dart/src/com/jetbrains/lang/dart/analyzer/DartAnalysisServerService.java:
```diff
@@ -657,6 +661,35 @@ public void run() {
+  @NotNull
+  public List<HoverInformation> analysis_getHover(@NotNull final String _filePath, final int offset) {
+    final String filePath = FileUtil.toSystemDependentName(_filePath);
+    final AtomicReference<List<HoverInformation>> result = new AtomicReference<List<HoverInformation>>();
```
You do not use 'atomic' features of this reference. That's correct, we do not need them, we are not afraid of a race here. So simple `Ref` that we use in all other methods here is enough.

alexander-doroshko commented on Dart/src/com/jetbrains/lang/dart/analyzer/DartAnalysisServerService.java:
```diff
@@ -657,6 +661,35 @@ public void run() {
     });
   }
 
+  @NotNull
```
It returns null if timed out

alexander-doroshko commented on Dart/src/com/jetbrains/lang/dart/analyzer/DartAnalysisServerService.java:
```diff
@@ -657,6 +661,35 @@ public void run() {
+    final CountDownLatch latch = new CountDownLatch(1);
+    server.analysis_getHover(filePath, offset, new GetHoverConsumer() {
+      @Override
+      public void computedHovers(HoverInformation[] hovers) {
```
In what cases this array is longer than 1? Multi resolve like one part file for several main files?

scheglov replied:
Correct.
I don't know for sure if DAS actually ever returns more than one, but it could.

alexander-doroshko commented on Dart/src/com/jetbrains/lang/dart/ide/documentation/DartDocUtil.java:
```diff
@@ -268,7 +269,7 @@ private static void appendClassSignature(final StringBuilder builder, final Dart
     final DartType superClass = dartClass.getSuperClass();
     if (superClass != null) {
-      builder.append("<br/>extends ").append(StringUtil.escapeXml(superClass.getText()));
+      builder.append(" extends ").append(StringUtil.escapeXml(superClass.getText()));
```
Don't you like it on separate lines like done for Java? Implements list may be long.

alexander-doroshko commented on Dart/src/com/jetbrains/lang/dart/ide/documentation/DartDocumentationProvider.java:
```diff
@@ -94,18 +98,89 @@ else if (componentType == DartComponentType.CLASS) {
+      builder.append("<br>");
+      if (containingLibraryName != null) {
+        builder.append("<b>Containing library:</b> ");
+        builder.append(StringUtil.escapeXml(containingLibraryName));
```
lib name may be an empty string

alexander-doroshko commented on Dart/src/com/jetbrains/lang/dart/ide/documentation/DartDocumentationProvider.java:
```diff
@@ -94,18 +98,89 @@ else if (componentType == DartComponentType.CLASS) {
-  public String generateDoc(PsiElement element, PsiElement originalElement) {
-    return DartDocUtil.generateDoc(element);
+  @Nullable
+  private static String generateDocServer(PsiElement element) {
```
Idea: unify doc preview regardless of the source of information: server or IDE. Extract a method that will build resulting HTML based on prepared input data.

alexander-doroshko commented:
Need a test for server based docs. At least one for each kind of declaration.
Review completed.

alexander-doroshko commented:
WEB-15936 is about hover, it is a bit more tricky to fix than docs.

> Note that even the static type is not always the same as the declared type because of the type promotion process or "strong mode".

I'm afraid I'm out of context. Can you provide an example?

scheglov commented:
This is an example of type promotion.

See `19.1.1 Type Promotion` and `17.5 If` in the Dart specification.

```
class A {}
class B extends A {
  bbb() {}
}

main(A p) {
  if (p is B) {
    p.bbb();
  }
}
```

  In this example the declared type of `p` is `A`.
  But the the static type is `B`, because we "promoted" the type from the `if` statement.
![image](https://cloud.githubusercontent.com/assets/384794/11749788/93a7a586-9fe4-11e5-98db-ce256abdef50.png)

scheglov commented:
We can also apply `type propagation` to a variable - additional changes to the type, which are not mandated by the specification. The following code has a required warning, but it has all three types - declared, static and propagated.

```
class A {}
class B extends A {
  bbb() {}
}
class C extends B {
  ccc() {}
}

main(A p) {
  if (p is B) {
    p.bbb();
    if (p is! C) {
      return;
    }
    p.ccc();
  }
}
```

![image](https://cloud.githubusercontent.com/assets/384794/11750033/02e9fcfe-9fe6-11e5-8ef0-e8266b9ddb3c.png)

alexander-doroshko commented:
Oh, too many types: declared, static, promoted, propagated, inferred. Is promoted == static? Is propagated == inferred?

scheglov commented:
Yes, too many types... of types :-)
It depends on the contexts and even options.

Yes, when we do `type promotion`, we update the `static` type.

In the strong mode static types of some parameters and return types of methods are replaced. This is called `inferred` types internally, we AFAIK we expose them as `static` types.

Types propagation is for local variables and parameters. No, `propagated != inferred`.

scheglov commented:
The PR is updated.

1. I agree that signature for classes is not as useful as it could be. This [was fixed](https://github.com/dart-lang/sdk/commit/168556a3e44364ac02555b4b3aec3d5821fb1545) in DAS.

2. Removed unused imports.

3. Removed usage of `Ref` at all, just `List` now, always not `null`, but maybe empty.

4. Yes, `extends` and `implements` can be both long. But not always. And they also can be long even if we put them on separate lines. So, I'd like to keep in on a single line and let the wrapper to do its work as needed.

5. Yes, library names can be empty. I converted all the checks to `!StringUtil.isEmpty()`.

6. Implemented a common documentation generator using given parameters.

7. Ported with some changes tests.
